rpc/rpc.go

```diff
@@ -137,83 +125,123 @@ type ErrorReporter interface {
 // Once a connection is created, it will immediately start receiving
 // requests from the transport.
 func NewConn(t Transport, opts *Options) *Conn {
-	bgctx, bgcancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// We use an errgroup to link the lifetime of background tasks
+	// to each other.
+	g, ctx := errgroup.WithContext(ctx)
+
 	c := &Conn{
 		transport: t,
 		shut:      make(chan struct{}),
-		bgctx:     bgctx,
-		bgcancel:  bgcancel,
+		bgctx:     ctx,
+		bgcancel:  cancel,
 		answers:   make(map[answerID]*answer),
 		imports:   make(map[importID]*impent),
+		sendq:     mpsc.New(),
 	}
 	if opts != nil {
 		c.bootstrap = opts.BootstrapClient
-		c.reporter = opts.ErrorReporter
+		c.er.ErrorReporter = opts.ErrorReporter
 		c.abortTimeout = opts.AbortTimeout
 	}
 	if c.abortTimeout == 0 {
 		c.abortTimeout = 100 * time.Millisecond
 	}
-	c.tasks.Add(1)
+
+	// start background tasks
+	g.Go(c.backgroundTask(c.send))
+	g.Go(c.backgroundTask(c.receive))
+
+	// monitor background tasks
 	go func() {
-		abortErr := c.receive(c.bgctx)
-		c.tasks.Done()
+		err := g.Wait()
```

Comment:
Did you mean to swap the Done() for a Wait()? My hunch is this is the source of your deadlock.

Reply:
I think that's correct, but you've put me on the right track.  I think the issue is that the `errgroup.Group`'s context returns on the first non-nil error, so I need to handle the `context.Canceled` here instead of in `c.backgroundTask`.

Testing now.

Reply:
No dice.  The `Done()` you're looking for is in `backgroundTask`.  This just waits until either `c.send` or `c.recv` terminate.